// @parity/jsonrpc/lib/index.js
'use strict'; // Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

var db = require('./interfaces/db');
var eth = require('./interfaces/eth');
var net = require('./interfaces/net');
var parity = require('./interfaces/parity');
var personal = require('./interfaces/personal');
var shh = require('./interfaces/shh');
var signer = require('./interfaces/signer');
var trace = require('./interfaces/trace');
var web3 = require('./interfaces/web3');

module.exports = {
  db: db,
  eth: eth,
  parity: parity,
  net: net,
  personal: personal,
  shh: shh,
  signer: signer,
  trace: trace,
  web3: web3 };
